api/context.go

```diff
@@ -97,41 +110,95 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		// attempt to parse the session token from the header
+		if ah := r.Header.Get(model.HEADER_AUTH); ah != "" {
+			if len(ah) > 6 && strings.ToUpper(ah[0:6]) == model.HEADER_BEARER {
+				sessionId = ah[7:]
+			}
 		}
-	}
 
-	// attempt to parse the session token from the cookie
-	if sessionId == "" {
-		if cookie, err := r.Cookie(model.SESSION_TOKEN); err == nil {
-			sessionId = cookie.Value
+		// attempt to parse the session token from the cookie
+		if sessionId == "" {
+			if cookie, err := r.Cookie(model.SESSION_TOKEN); err == nil {
+				sessionId = cookie.Value
+			}
 		}
-	}
 
-	if sessionId != "" {
+		if sessionId != "" {
+
+			var session *model.Session
+			if ts, ok := sessionCache.Get(sessionId); ok {
+				session = ts.(*model.Session)
+			}
+
+			if session == nil {
+				if sessionResult := <-Srv.Store.Session().Get(sessionId); sessionResult.Err != nil {
+					c.LogError(model.NewAppError("ServeHTTP", "Invalid session", "id="+sessionId+", err="+sessionResult.Err.DetailedError))
+				} else {
+					session = sessionResult.Data.(*model.Session)
+				}
+			}
 
-		var session *model.Session
-		if ts, ok := sessionCache.Get(sessionId); ok {
-			session = ts.(*model.Session)
+			if session == nil || session.IsExpired() {
+				c.RemoveSessionCookie(w)
+				c.Err = model.NewAppError("ServeHTTP", "Invalid or expired session, please login again.", "id="+sessionId)
+				c.Err.StatusCode = http.StatusUnauthorized
+			} else {
+				c.Session = *session
+			}
+		}
+	} else {
+		var accessData *model.AccessData
+		if td, ok := accessTokenCache.Get(token); ok {
+			accessData = td.(*model.AccessData)
 		}
 
-		if session == nil {
-			if sessionResult := <-Srv.Store.Session().Get(sessionId); sessionResult.Err != nil {
-				c.LogError(model.NewAppError("ServeHTTP", "Invalid session", "id="+sessionId+", err="+sessionResult.Err.DetailedError))
+		if accessData == nil {
+			if result := <-Srv.Store.AccessData().Get(token); result.Err != nil {
+				c.LogError(model.NewAppError("ServeHTTP", "Invalid access token", "err="+result.Err.DetailedError))
 			} else {
-				session = sessionResult.Data.(*model.Session)
+				accessData = result.Data.(*model.AccessData)
 			}
 		}
 
-		if session == nil || session.IsExpired() {
-			c.RemoveSessionCookie(w)
-			c.Err = model.NewAppError("ServeHTTP", "Invalid or expired session, please login again.", "id="+sessionId)
+		if accessData == nil {
+			c.Err = model.NewAppError("ServeHTTP", "Invalid access token", "")
+			c.Err.StatusCode = http.StatusUnauthorized
+		} else if accessData.IsExpired() {
+			// Remove access token, related auth code from DB as well as from cache
+			if err := RevokeAccessToken(token); err != nil {
+				l4g.Error("Encountered an error revoking an access token, err=" + err.Message)
+			}
+
+			c.Err = model.NewAppError("ServeHTTP", "Access token has expired", "")
 			c.Err.StatusCode = http.StatusUnauthorized
 		} else {
-			c.Session = *session
+			var session *model.Session
```

Comment:
Yeah we might be able to do that, I just figured that since it's such critical path stuff that it might be better to have it split out explicitly. 0/5 as well